specifications/FiniteMonotonic/BlogPost.md

```diff
@@ -0,0 +1,419 @@
+The ability to explore all possible interleavings of events makes concurrency simple to reason about.
+For this TLA⁺ uses something called finite model-checking, which is really just a breadth-first search through the entire state space.
+The key here - and this really must be emphasized - is that the model is *finite*.
+There can't be an infinite number of states, or of course the model checker will run forever.
+
+Unfortunately, a lot of distributed systems are designed with infinitely-expanding state at their core.
+You want your system to be *monotonic*, which means it never revisits an earlier state.
+This is because ordering things in a distributed system is very difficult.
+If your system always progresses through states, never returning to earlier ones, it gives you a simple way to order your system and say it was in one state before it was in another state.
+A whole class of problems evaporate.
+
+Here is an example: a number that only increases in value, which we'll call a monotonic counter.
+Monotonic counters are used everywhere in distributed systems.
+In fact, I struggle to think of a distributed system that *doesn't* have a monotonic counter somewhere in its design.
+Here's a short list of ways I've seen monotonic counters used:
+ * transaction IDs
+ * live configuration versioning
+ * indices in an append-only log
+ * proposal ballot numbers in consensus algorithms
+
+So here's the problem: we want to subject distributed systems to finite modeling, but they try to cram as much infinity as possible into their system designs.
+How do we square this circle?
+
+## The hammer: state restrictions
+
+Here's a method long used by many TLA⁺ practitioners, including myself.
+Just pick an arbitrary point and say "okay, we've explored enough of the state space, don't go beyond here".
+So your monotonic counters are limited to, say, 10 or below.
+If you haven't found a bug by the time your counter hits 10 it probably doesn't exist, so the thinking goes.
+This is actually pretty good thinking!
+Indeed, when designing a model you're often making tradeoffs between state exploration and your model finishing before the heat death of the universe.
+It's rare, for example, for a TLA⁺ model to use more than three nodes in a distributed system despite production systems often involving many, many more nodes than that.
+TLA⁺ is still extraordinarily effective at finding bugs in these models, which is a testament to the complexity of concurrent & distributed systems generally.
+Us programmers can barely think straight about one computer, let alone three or more!
+
+Unfortunately, if you add state restrictions you've cut yourself off from many powerful TLA⁺ features (see page 247, section 14.3.5 of [*Specifying Systems*](https://lamport.azurewebsites.net/tla/book.html) for details).
+In particular, you'll lose the ability to analyze liveness (good things eventually happen) or abstraction & refinement (showing that one spec is related to another).
```

Comment:
This is too broad because checking the safety part of refinement `LLInit /\ [][LLNext]_v => HLInit /\ [][HLNext]_v` works just fine.

Reply:
Isn't there a large problem with success being reported due to vacuous truth? Or is my understanding wrong; related https://github.com/tlaplus/tlaplus/issues/839

Reply:
Please elaborate.

Reply:
I mean when checking liveness with state restrictions. It doesn't feel particularly safe. Also where is the formula `LLInit /\ [][LLNext]_v => HLInit /\ [][HLNext]_v` coming from?

Reply:
LLInit and LLNext are the initial predicate and next-state relation of the low-level spec and HLInit and HLNext are the ones from the high-level/refined spec.